File: modernized-python/backend/app/services/order_service.py
"""
Order business logic. This is the direct Python counterpart of
OrderService.java on the legacy side -- same rules, same rounding, same
status workflow. See MIGRATION.md at the repo root for the side-by-side
mapping and why the numbers below must match exactly.
"""
from decimal import Decimal, ROUND_HALF_UP

from sqlalchemy.orm import Session

from app.models import OrderLineItem, Product, PurchaseOrder

BULK_DISCOUNT_THRESHOLD = 10
BULK_DISCOUNT_RATE = Decimal("0.10")
TAX_RATE = Decimal("0.08")

TWO_PLACES = Decimal("0.01")


class BusinessRuleError(Exception):
    """Mirrors BusinessRuleException on the legacy side."""


def _round(value: Decimal) -> Decimal:
    return value.quantize(TWO_PLACES, rounding=ROUND_HALF_UP)


def calculate_line_discount(quantity: int, line_subtotal: Decimal) -> Decimal:
    """Bulk-discount rule: quantity >= BULK_DISCOUNT_THRESHOLD earns
    BULK_DISCOUNT_RATE off that line's subtotal."""
    if quantity >= BULK_DISCOUNT_THRESHOLD:
        return _round(line_subtotal * BULK_DISCOUNT_RATE)
    return _round(Decimal("0"))


def create_order(db: Session, customer_id: int) -> PurchaseOrder:
    order = PurchaseOrder(customer_id=customer_id, status=PurchaseOrder.STATUS_DRAFT)
    db.add(order)
    db.commit()
    db.refresh(order)
    return order


def add_line_item(db: Session, order_id: int, product_id: int, quantity: int) -> PurchaseOrder:
    order = db.get(PurchaseOrder, order_id)
    if order is None:
        raise BusinessRuleError(f"Order {order_id} does not exist.")
    if order.status != PurchaseOrder.STATUS_DRAFT:
        raise BusinessRuleError("Line items can only be added while an order is in DRAFT status.")
    if quantity <= 0:
        raise BusinessRuleError("Quantity must be a positive number.")

    product = db.get(Product, product_id)
    if product is None:
        raise BusinessRuleError(f"Product {product_id} does not exist.")

    line_subtotal = _round(product.unit_price * quantity)
    line_discount = calculate_line_discount(quantity, line_subtotal)
    line_total = _round(line_subtotal - line_discount)

    line_item = OrderLineItem(
        order_id=order.id,
        product_id=product.id,
        quantity=quantity,
        unit_price=product.unit_price,
        line_subtotal=line_subtotal,
        line_discount=line_discount,
        line_total=line_total,
    )
    db.add(line_item)
    db.commit()

    _recalculate_order_totals(db, order_id)
    db.refresh(order)
    return order


def _recalculate_order_totals(db: Session, order_id: int) -> None:
    order = db.get(PurchaseOrder, order_id)

    subtotal = sum((item.line_subtotal for item in order.line_items), Decimal("0"))
    discount = sum((item.line_discount for item in order.line_items), Decimal("0"))

    taxable = subtotal - discount
    tax = _round(taxable * TAX_RATE)
    total = _round(taxable + tax)

    order.subtotal = _round(subtotal)
    order.discount_amount = _round(discount)
    order.tax_amount = tax
    order.total = total
    db.commit()


def submit_order(db: Session, order_id: int) -> PurchaseOrder:
    order = db.get(PurchaseOrder, order_id)
    if order is None:
        raise BusinessRuleError(f"Order {order_id} does not exist.")
    if order.status != PurchaseOrder.STATUS_DRAFT:
        raise BusinessRuleError("Only a DRAFT order can be submitted.")
    if not order.line_items:
        raise BusinessRuleError("Cannot submit an order with no line items.")

    order.status = PurchaseOrder.STATUS_SUBMITTED
    db.commit()
    db.refresh(order)
    return order


def _transition_from_submitted(db: Session, order_id: int, target_status: str) -> PurchaseOrder:
    order = db.get(PurchaseOrder, order_id)
    if order is None:
        raise BusinessRuleError(f"Order {order_id} does not exist.")
    if order.status != PurchaseOrder.STATUS_SUBMITTED:
        raise BusinessRuleError("Only a SUBMITTED order can be approved or rejected.")

    order.status = target_status
    db.commit()
    db.refresh(order)
    return order


def approve_order(db: Session, order_id: int) -> PurchaseOrder:
    return _transition_from_submitted(db, order_id, PurchaseOrder.STATUS_APPROVED)


def reject_order(db: Session, order_id: int) -> PurchaseOrder:
    return _transition_from_submitted(db, order_id, PurchaseOrder.STATUS_REJECTED)


def get_order(db: Session, order_id: int) -> PurchaseOrder | None:
    return db.get(PurchaseOrder, order_id)


def list_orders(db: Session) -> list[PurchaseOrder]:
    return db.query(PurchaseOrder).order_by(PurchaseOrder.id.desc()).all()

File: legacy-java/src/main/java/com/orderco/legacy/service/OrderService.java
package com.orderco.legacy.service;

import com.orderco.legacy.dao.ProductDao;
import com.orderco.legacy.dao.PurchaseOrderDao;
import com.orderco.legacy.model.Product;
import com.orderco.legacy.model.PurchaseOrder;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.sql.SQLException;
import java.util.List;

/**
 * Order business logic, isolated from both the web tier (Struts Actions)
 * and the persistence tier (DAOs) -- the same separation of concerns an EJB
 * session bean would enforce, just as a plain Java class instead of a
 * container-managed bean. See the top-level README for why this demo
 * doesn't stand up a full Java EE application server to host a literal
 * @Stateless session bean.
 *
 * Business rules implemented here (and re-implemented identically on the
 * modernized side -- see MIGRATION.md):
 *   - A line item with quantity >= BULK_DISCOUNT_THRESHOLD gets a
 *     BULK_DISCOUNT_RATE discount on that line's subtotal.
 *   - Tax is a flat TAX_RATE applied to (subtotal - discount).
 *   - Orders can only be submitted from DRAFT, and only approved/rejected
 *     from SUBMITTED.
 */
public class OrderService {

    public static final int BULK_DISCOUNT_THRESHOLD = 10;
    public static final BigDecimal BULK_DISCOUNT_RATE = new BigDecimal("0.10");
    public static final BigDecimal TAX_RATE = new BigDecimal("0.08");

    private final PurchaseOrderDao orderDao = new PurchaseOrderDao();
    private final ProductDao productDao = new ProductDao();

    public long createOrder(long customerId) throws SQLException {
        return orderDao.insertOrder(customerId);
    }

    public void addLineItem(long orderId, long productId, int quantity) throws SQLException, BusinessRuleException {
        PurchaseOrder order = orderDao.findById(orderId);
        if (order == null) {
            throw new BusinessRuleException("Order " + orderId + " does not exist.");
        }
        if (!PurchaseOrder.STATUS_DRAFT.equals(order.getStatus())) {
            throw new BusinessRuleException("Line items can only be added while an order is in DRAFT status.");
        }
        if (quantity <= 0) {
            throw new BusinessRuleException("Quantity must be a positive number.");
        }

        Product product = productDao.findById(productId);
        if (product == null) {
            throw new BusinessRuleException("Product " + productId + " does not exist.");
        }

        BigDecimal lineSubtotal = product.getUnitPrice().multiply(BigDecimal.valueOf(quantity));
        BigDecimal lineDiscount = calculateLineDiscount(quantity, lineSubtotal);
        BigDecimal lineTotal = lineSubtotal.subtract(lineDiscount).setScale(2, RoundingMode.HALF_UP);

        orderDao.insertLineItem(orderId, productId, quantity, product.getUnitPrice(), lineSubtotal, lineDiscount, lineTotal);
        recalculateOrderTotals(orderId);
    }

    /**
     * Bulk-discount rule: quantity >= BULK_DISCOUNT_THRESHOLD earns
     * BULK_DISCOUNT_RATE off that line's subtotal.
     */
    BigDecimal calculateLineDiscount(int quantity, BigDecimal lineSubtotal) {
        if (quantity >= BULK_DISCOUNT_THRESHOLD) {
            return lineSubtotal.multiply(BULK_DISCOUNT_RATE).setScale(2, RoundingMode.HALF_UP);
        }
        return BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP);
    }

    private void recalculateOrderTotals(long orderId) throws SQLException {
        PurchaseOrder order = orderDao.findById(orderId);

        BigDecimal subtotal = BigDecimal.ZERO;
        BigDecimal discount = BigDecimal.ZERO;
        for (int i = 0; i < order.getLineItems().size(); i++) {
            subtotal = subtotal.add(order.getLineItems().get(i).getLineSubtotal());
            discount = discount.add(order.getLineItems().get(i).getLineDiscount());
        }

        BigDecimal taxable = subtotal.subtract(discount);
        BigDecimal tax = taxable.multiply(TAX_RATE).setScale(2, RoundingMode.HALF_UP);
        BigDecimal total = taxable.add(tax).setScale(2, RoundingMode.HALF_UP);

        orderDao.updateOrderTotals(orderId, subtotal.setScale(2, RoundingMode.HALF_UP),
                discount.setScale(2, RoundingMode.HALF_UP), tax, total);
    }

    public void submitOrder(long orderId) throws SQLException, BusinessRuleException {
        PurchaseOrder order = orderDao.findById(orderId);
        if (order == null) {
            throw new BusinessRuleException("Order " + orderId + " does not exist.");
        }
        if (!PurchaseOrder.STATUS_DRAFT.equals(order.getStatus())) {
            throw new BusinessRuleException("Only a DRAFT order can be submitted.");
        }
        if (order.getLineItems().isEmpty()) {
            throw new BusinessRuleException("Cannot submit an order with no line items.");
        }
        orderDao.updateStatus(orderId, PurchaseOrder.STATUS_SUBMITTED);
    }

    public void approveOrder(long orderId) throws SQLException, BusinessRuleException {
        transitionFromSubmitted(orderId, PurchaseOrder.STATUS_APPROVED);
    }

    public void rejectOrder(long orderId) throws SQLException, BusinessRuleException {
        transitionFromSubmitted(orderId, PurchaseOrder.STATUS_REJECTED);
    }

    private void transitionFromSubmitted(long orderId, String targetStatus) throws SQLException, BusinessRuleException {
        PurchaseOrder order = orderDao.findById(orderId);
        if (order == null) {
            throw new BusinessRuleException("Order " + orderId + " does not exist.");
        }
        if (!PurchaseOrder.STATUS_SUBMITTED.equals(order.getStatus())) {
            throw new BusinessRuleException("Only a SUBMITTED order can be approved or rejected.");
        }
        orderDao.updateStatus(orderId, targetStatus);
    }

    public PurchaseOrder getOrder(long orderId) throws SQLException {
        return orderDao.findById(orderId);
    }

    public List<PurchaseOrder> listOrders() throws SQLException {
        return orderDao.findAll();
    }
}
